### ISSUE TITLE
WinRM service restarts during active connection, Packer shows exit code 0

### ISSUE THREAD

**Issue by ccoager** [hashicorp/packer#9433]
#### Overview of the Issue
I having an ongoing, reproducible issue with Packer and WinRM on Windows 2019.    The upload of the script always works but the execution of the script doesn't complete.  The WinRM service is restarting and Packer is returning exit code 0 even though the command doesn't complete.  Packer continues on with the next piece of code when it should be failing or retrying the command that last failed.

#### Reproduction Steps
Restart WinRM service during active connection for Packer between the host and VM.

### Packer version
1.5.4

### Simplified Packer Buildfile
Here is a snippet example of one of the commands.  This part seems irrelevant because it can happen with any command.

>     {
>       "type": "powershell",
>       "inline": [
>         "$temp = \"C:\\windows\\temp\"",
>         "Write-Host \"Stage create CFEngine ppkey\"",
>         "$p = start-process -FilePath \"cf-key\" -ArgumentList \"-Iv\" -Wait -NoNewWindow",
>         "$p.WaitForExit()",
>         "if (-Not (Test-Path \"C:\\Program Files\\Cfengine\\ppkeys\\localhost.pub\")) { Exit 1 }",
>         "if (-Not (Test-Path \"C:\\Program Files\\Cfengine\\ppkeys\\localhost.priv\")) { Exit 1 }",
>         "$p = start-process -FilePath \"cf-key\" -ArgumentList \"-p \"\"$env:programfiles\\cfengine\\ppkeys\\localhost.pub\"\"\" -Wait -NoNewWindow",
>         "$p.WaitForExit()",
>         "New-Item \"$temp\\stage_ppkey\" -ItemType file"
>       ]
>     },
> 

### Operating system and Environment details
Windows 2019 build 17763

### Log Fragments and crash.log files
In the example below, Packer tries executing the script at 13:59:49 and completes at 14:00:00 but the script didn’t complete.  If you look at the host logs below, you can see the it starting to execute it’s command at 1:59:59 then the WinRM restarts itself at 2:00:00.  Why is WinRM restarting itself?  This WinRM issue is reproducible and I don’t understand why it’s happening.  Also, Packer doesn’t detect that this step failed and keeps going on with further steps.  This produces an incomplete template and I am not able to detect the steps that failed.
[Packer and Event logs](https://gist.github.com/ccoager/13886f3a674f03edd68c723304681afe)

**Comment by nywilken** [comment]
Hi @ccoager thanks for reaching out. So two things here: 


One it looks like the call to CF engine triggers as expected. But it is most likely forking into a separate process that returns a success once forked so Packer believes the command executed without issue. We've seen this issue with other configuration management tools such as puppet via a shell or powershell provisioner. 


Two the WinRM resetting issue. Quickly glancing at the provided script I can't make out how running it would affect the WinRM connection. Nor do I have experience with CF engine to identify the possible issue.

Have you tried running the Packer build without the Powershell provisioner to see if the WinRM restart occurs? 

 If you haven't already I would replace the powershell script with a simpler script just to see if the WinRM connection restarts. If it does than we can start looking at that. If it doesn't then you may want to dive into the CF engine process to see if it is changing anything with the secure settings for WinRM. I hope this helps.

**Comment by ccoager** [comment]
@nywilken 
The "$p.WaitForExit()" instructs the process component to wait indefinitely for the associated process to exit, according to the documentation.  That should cover the case of waiting for the parent/child to exit cleanly.

WinRM resetting seems to be specific to Windows 2019.  The same code seems to run fine on 2012 and 2016.  The only other choice I see is Windows shell which still uses WinRM.  I can try testing it.

The issue with Packer I am curious about, the WinRM service restarted and killed the Powershell provisioner mid process.  Why is Packer reporting exit 0?  I don't think Packer should report no issue in this scenario.  Do you agree?

**Comment by nywilken** [comment]
> The issue with Packer I am curious about, the WinRM service restarted and killed the Powershell provisioner mid process. Why is Packer reporting exit 0? I don't think Packer should report no issue in this scenario. Do you agree?

Hi @ccoager apologies for the slow response here. I agree. There is actually an open issue regarding Powershell report exit 0 when there are errors https://github.com/hashicorp/packer/issues/4916

The issue stems from Packer returning $LastExitCode which is only relevant for native Windows applications and not a native PowerShell script exit code. Which may be the issue here regarding the exit code. If the WinRM connection is disconnected Packer will try to reconnect, this is handled outside of the provisioner, so is it possible that the provisioner is getting an incorrect response back :thinking: 

If possible could you share your 2019 packer template I would like to test against this to see if there is anything we might be able to do recognize this issue better.

Does the restarting of WinRM on 2019 only happen after the CFEngine provisioner execution? Have you checked for any known issues? 

Cheers!

**Comment by ccoager** [comment]
@nywilken I figured out what the issue was.  It turned out be a user_data_file script to configure WinRM was continuing to run after reboots.  The user data script was configuring WinRM so one of the steps was resetting the service while the Powershell provisioner was running.  Since I already baked the WinRM config in our starting point templates, I just removed the user data script.

Since you have another ticket open for the Powershell exit codes, it is safe to close this one now.  Thanks for the help!
